### PR TITLE
avm2: Remove or validate some uses of coerce_to_object

### PR DESCRIPTION
**Pull request by Lord-McSweeney**
Progresses #9129, which can now get into the game, though errors are thrown.

### CODE REVIEW

File: core/src/avm2/activation.rs [hunk 1]
```diff
@@ -2898,34 +2898,62 @@ impl<'a, 'gc> Activation<'a, 'gc> {
...
-            .and_then(|c| c.as_class_object())
-            .ok_or("Cannot coerce a value to a type that is null, undefined, or not a class")?;
-        let value = self.pop_stack();
+        let class = self.pop_stack();
 
-        if value.is_of_type(self, class) {
-            self.push_stack(value);
-        } else {
-            self.push_stack(Value::Null);
+        if matches!(class, Value::Undefined) {
```
Comment (adrian17) [inline core/src/avm2/activation.rs:2903]: can we check for `Value::Null | Value::Undefined` in one go here?

File: core/src/avm2/activation.rs [hunk 2]
```diff
@@ -2172,7 +2172,7 @@ impl<'a, 'gc> Activation<'a, 'gc> {
     fn op_check_filter(&mut self) -> Result<FrameControl<'gc>, Error<'gc>> {
         let xml = self.avm2().classes().xml;
         let xml_list = self.avm2().classes().xml_list;
-        let value = self.pop_stack().coerce_to_object(self)?;
+        let value = self.pop_stack().coerce_to_object_or_typeerror(self, None)?;
```
[the remark this answers is not in the document]
Comment (Lord-McSweeney) [reply core/src/avm2/activation.rs:2175]: I checked and it does throw the error for null/undefined values.

File: core/src/avm2/activation.rs (hunk 1 above)
Comment (Lord-McSweeney) [reply core/src/avm2/activation.rs:2903]: `undefined` throws the undefined error, while `null` and all other primitives throw the null error.

File: core/src/avm2/globals/array.rs [hunk 3]
```diff
@@ -182,7 +182,10 @@ pub fn concat<'gc>(
         .unwrap_or_else(|| ArrayStorage::new(0));
 
     for arg in args {
-        if let Some(other_array) = arg.coerce_to_object(activation)?.as_array_storage() {
+        if let Some(other_array) = arg
+            .as_object()
+            .and_then(|obj| obj.as_array_storage().map(|a| a.clone()))
```
[the remark this answers is not in the document]
Comment (Lord-McSweeney) [reply core/src/avm2/globals/array.rs:187]: If I remove the clone I get
```
error[E0515]: cannot return reference to function parameter `obj`
   --> core/src/avm2/globals/array.rs:187:29
    |
187 |             .and_then(|obj| obj.as_array_storage())
    |                             ^^^^^^^^^^^^^^^^^^^^^^ returns a reference to data owned by the current function
```
Is there another way to fix that without having a nested
```
if let Some(...) = ... {
    if let Some(...) = ... {
        ...
    }
}
```

File: core/src/avm2/activation.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (adrian17) [reply core/src/avm2/activation.rs:2175]: OK, I just rechecked, sorry.

File: core/src/avm2/globals/array.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Aaron1011) [reply core/src/avm2/globals/array.rs:187]: Use `.as_object().as_ref()`

File: core/src/avm2/activation.rs (hunk 1 above)
Comment (adrian17) [reply core/src/avm2/activation.rs:2903]: Hmm, ok, I see it; kinda weird ;D
Anyway, if I wrote it, I'd still check for `Null|Undefined` as a pattern, but if you don't wanna, that's okay too.

**Comment by Aaron1011** [comment]
Can you add a test?

**Review by Aaron1011 — APPROVED** [review]

**Pull request merged**